### PR TITLE
IMT6: Proper DelegatingTimelockService

### PR DESCRIPTION
**Pull request by gsheasby**
**Goals (and why)**: Make InMemoryTimeLockServices use the same delegations as RemoteTimeLockServiceAdapter, so that we can reasonably test lock watches.

**Implementation Description (bullets)**: Bunch of wiring, fixed some tests

**Testing (What was existing testing like?  What have you done to improve it?)**: Added consistency test in InMemoryTimeLockServicesTest (after an initial version broke this and was a pain to root-cause); changed many tests in SerializableTransactionTest so that they use the serializable table. Much less yolo-creation of transaction objects.

**Concerns (what feedback would you like?)**: Is DelegatingTimelockService sufficiently faithful? What else needs to be changed for this workstream to be done?

**Where should we start reviewing?**: Comparing DefaultLockAndTimestampServices and TimeLockHelperServices (to check the creation stuff is all fine), and then seeing how it's used in InMemoryTimeLockServices.

**Priority (whenever / two weeks / yesterday)**: this week

### CODE REVIEW

File: atlasdb-config/src/main/java/com/palantir/atlasdb/factory/DefaultLockAndTimestampServiceFactory.java [hunk 1]
```diff
@@ -341,12 +338,18 @@ private static LockAndTimestampServices getLockAndTimestampServices(
...
-        LockWatchManagerInternal lockWatchManager =
-                LockWatchManagerImpl.create(metricsManager, schemas, lockWatchingService, cachingConfig);
-
         Supplier<InternalMultiClientConjureTimelockService> multiClientTimelockServiceSupplier =
                 getMultiClientTimelockServiceSupplier(serviceProvider);
 
+        Supplier<Optional<RequestBatchersFactory.MultiClientRequestBatchers>> requestBatcherProvider =
+                () -> timelockRequestBatcherProviders.map(batcherProviders -> ImmutableMultiClientRequestBatchers.of(
+                        batcherProviders.commitTimestamps().getBatcher(multiClientTimelockServiceSupplier),
+                        batcherProviders.startTransactions().getBatcher(multiClientTimelockServiceSupplier)));
```
Comment (Jolyon-S) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/factory/DefaultLockAndTimestampServiceFactory.java:347]: Thoughts on pushing this into `TimelockHelperServices` too? It's only used in there, and we could just pass in `timelockRequestBatcherProviders` instead.

File: atlasdb-tests-shared/src/main/java/com/palantir/atlasdb/transaction/impl/AbstractSerializableTransactionTest.java [hunk 2]
```diff
@@ -130,7 +131,7 @@ protected TransactionManager createManager() {
 
     @Override
     protected Transaction startTransaction() {
-        return startTransactionWithOptions(new TransactionOptions());
+        return Iterables.getOnlyElement(txMgr.startTransactions(List.of(PreCommitConditions.NO_OP)));
```
Comment (Jolyon-S) [inline atlasdb-tests-shared/src/main/java/com/palantir/atlasdb/transaction/impl/AbstractSerializableTransactionTest.java:134]: I get that you're making this class less terrible, but is it directly tied to the other changes? Maybe it is (because we're now _actually_ using a serializable table), but just wanted to check.

File: atlasdb-tests-shared/src/main/java/com/palantir/atlasdb/transaction/impl/AbstractTransactionTest.java [hunk 3]
```diff
@@ -126,33 +113,7 @@ protected boolean supportsReverse() {
...
-                timestampCache,
-                GET_RANGES_EXECUTOR,
-                DEFAULT_GET_RANGES_CONCURRENCY,
-                MultiTableSweepQueueWriter.NO_OP,
-                MoreExecutors.newDirectExecutorService(),
-                true,
-                () -> TRANSACTION_CONFIG,
-                ConflictTracer.NO_OP,
-                new SimpleTableLevelMetricsController(metricsManager));
+        return Iterables.getOnlyElement(txMgr.startTransactions(List.of(PreCommitConditions.NO_OP)));
```
Comment (Jolyon-S) [inline atlasdb-tests-shared/src/main/java/com/palantir/atlasdb/transaction/impl/AbstractTransactionTest.java:116]: Thank the heavens

File: atlasdb-tests-shared/src/main/java/com/palantir/atlasdb/transaction/impl/TestTransactionManagerImpl.java [hunk 4]
```diff
@@ -175,36 +178,7 @@ public Transaction commitAndStartNewTransaction(Transaction tx) {
...
-                        getRangesExecutor,
-                        defaultGetRangesConcurrency,
-                        sweepQueueWriter,
-                        deleteExecutor,
-                        validateLocksOnReads,
-                        () -> TRANSACTION_CONFIG,
-                        ConflictTracer.NO_OP,
-                        tableLevelMetricsController),
-                pathTypeTracker);
+        return Iterables.getOnlyElement(startTransactions(ImmutableList.of(PreCommitConditions.NO_OP)));
```
Comment (Jolyon-S) [inline atlasdb-tests-shared/src/main/java/com/palantir/atlasdb/transaction/impl/TestTransactionManagerImpl.java:181]: this could almost be a default method on the `TestTransactionManager` interface (but fine here too).

Comment (Jolyon-S) [reply atlasdb-tests-shared/src/main/java/com/palantir/atlasdb/transaction/impl/TestTransactionManagerImpl.java:181]: Also we could do something similar to `commitAndStartNewTransaction`.

File: atlasdb-tests-shared/src/main/java/com/palantir/timelock/paxos/DelegatingTimelockService.java [hunk 5]
```diff
@@ -16,39 +16,53 @@
...
+            CommitTimestampGetter commitTimestampGetter) {
+        this.transactionStarter = transactionStarter;
+        this.lockLeaseService = lockLeaseService;
         this.timelock = timelock;
+        this.commitTimestampGetter = commitTimestampGetter;
+    }
+
+    @Override
+    public List<StartIdentifiedAtlasDbTransactionResponse> startIdentifiedAtlasDbTransactionBatch(int count) {
+        return transactionStarter.startIdentifiedAtlasDbTransactionBatch(count);
```
Comment (Jolyon-S) [inline atlasdb-tests-shared/src/main/java/com/palantir/timelock/paxos/DelegatingTimelockService.java:55]: does this make the default implementation of this method redundant? If so, can we remove it?

File: atlasdb-tests-shared/src/main/java/com/palantir/timelock/paxos/DelegatingTimelockService.java [hunk 6]
```diff
@@ -16,39 +16,53 @@
...
     }
 
     @Override
     public long getFreshTimestamp() {
-        return timelock.getFreshTimestamp();
+        return getFreshTimestamps(1).getLowerBound();
```
Comment (Jolyon-S) [inline atlasdb-tests-shared/src/main/java/com/palantir/timelock/paxos/DelegatingTimelockService.java:60]: I'm curious why this change was necessary - sure, it might be that under the hood now, but do we guarantee that we will never change `getFreshTimestamp`? Probably, but may as well have it go through the same codepath as the rest.

File: atlasdb-tests-shared/src/main/java/com/palantir/timelock/paxos/InMemoryTimelockServices.java [hunk 7]
```diff
@@ -126,6 +147,23 @@ protected void before() throws IOException {
...
+                delegate.getTimelockService(),
+                LockWatchCachingConfig.builder().build(),
+                Optional::empty);
+
+        RedirectRetryTargeter redirectRetryTargeter = timeLockAgent.redirectRetryTargeter();
+        ConjureTimelockService cts =
+                ConjureTimelockResource.jersey(redirectRetryTargeter, _unused -> delegate.getTimelockService());
+        NamespacedConjureTimelockService ncts = new NamespacedConjureTimelockServiceImpl(cts, client);
+        LeaderTimeGetter ltg = new LegacyLeaderTimeGetter(ncts);
+        lockLeaseService = LockLeaseService.create(ncts, ltg);
```
Comment (Jolyon-S) [inline atlasdb-tests-shared/src/main/java/com/palantir/timelock/paxos/InMemoryTimelockServices.java:164]: I appreciate that Atlas doesn't have a good record of variable names, but please can we bias towards using `timelockService` or `conjureTimelockService` over `cts`, and likewise?

File: atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java [hunk 8]
```diff
@@ -1203,8 +1206,9 @@ public void commitThrowsIfRolledBackAtCommitTime_expiredLocks() {
         PreCommitCondition condition =
                 unused -> doReturn(ImmutableSet.of()).when(timelockService).refreshLockLeases(any());
 
-        LockImmutableTimestampResponse res = timelockService.lockImmutableTimestamp();
-        long transactionTs = timelockService.getFreshTimestamp();
+        ConjureStartTransactionsResponse conjureResponse = startTransactionWithWatches();
+        LockImmutableTimestampResponse res = conjureResponse.getImmutableTimestamp();
+        long transactionTs = conjureResponse.getTimestamps().start();
```
Comment (Jolyon-S) [inline atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java:1211]: probably worth extracting a common method out for this

File: atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java [hunk 9]
```diff
@@ -1238,6 +1243,7 @@ public void commitThrowsIfRolledBackAtCommitTime_alreadyAborted() {
 
         snapshot.put(TABLE, ImmutableMap.of(cell, PtBytes.toBytes("value")));
 
+        Assertions.setMaxStackTraceElementsDisplayed(40);
```
Comment (Jolyon-S) [inline atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java:1246]: Probably worth adding a comment since this is pretty arcane.

File: atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java [hunk 10]
```diff
@@ -1269,7 +1277,6 @@ public void commitDoesNotThrowIfAlreadySuccessfullyCommitted() {
     public void validateLocksOnReadsIfThoroughlySwept() {
         long transactionTs = timelockService.getFreshTimestamp();
         LockImmutableTimestampResponse res = timelockService.lockImmutableTimestamp();
-
```
Comment (Jolyon-S) [inline atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java:1272]: ok

File: atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java [hunk 11]
```diff
@@ -2097,35 +2114,36 @@ private Transaction getSnapshotTransactionWith(
...
+                timestampCache,
+                getRangesExecutor,
+                defaultGetRangesConcurrency,
+                MultiTableSweepQueueWriter.NO_OP,
+                MoreExecutors.newDirectExecutorService(),
+                validateLocksOnReads,
+                () -> transactionConfig,
+                ConflictTracer.NO_OP,
+                tableLevelMetricsController);
+        // hack (?), tracking lock
```
Comment (Jolyon-S) [inline atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java:2144]: can you elaborate?

File: atlasdb-tests-shared/src/test/java/com/palantir/timelock/paxos/InMemoryTimelockServicesTest.java [hunk 12]
```diff
@@ -59,7 +62,31 @@ public void tearDown() {
...
+                .startTimestampAndPartition()
+                .timestamp();
+        testTs(ts3, ts4);
+
+        long ts5 = delegatingTimelockService.getFreshTimestamp();
+        testTs(ts4, ts5);
+    }
+
+    private void testTs(long ts2, long ts3) {
+        assertThat(ts2).isLessThan(ts3);
```
Comment (Jolyon-S) [inline atlasdb-tests-shared/src/test/java/com/palantir/timelock/paxos/InMemoryTimelockServicesTest.java:89]: Also for naming: this should probably be: `assertTimestampsAreOrdered` or something, with parameters and variables throughout `timestamp1`, `timestamp2` instead of `tsX`.

Comment (Jolyon-S) [reply atlasdb-tests-shared/src/test/java/com/palantir/timelock/paxos/InMemoryTimelockServicesTest.java:89]: Also, is there any reason why we can't make this stronger - fast forward the timestamp, and assert that they're the same?

Comment (Jolyon-S) [reply atlasdb-tests-shared/src/test/java/com/palantir/timelock/paxos/InMemoryTimelockServicesTest.java:89]: Maybe we don't want to commit to that (and only want to commit to the fact that they are monotonically increasing).

File: lock-api/src/main/java/com/palantir/lock/client/LockLeaseService.java [hunk 13]
```diff
@@ -42,7 +42,7 @@
 import java.util.UUID;
 import java.util.stream.Collectors;
 
-class LockLeaseService implements AutoCloseable {
+public class LockLeaseService implements AutoCloseable {
```
Comment (Jolyon-S) [inline lock-api/src/main/java/com/palantir/lock/client/LockLeaseService.java:45]: We have to be careful - if I'm not mistaken, making this public exposes more API, and also possibly exposes us to ABI breaks if the user should ever decide to do something naughty. It's possibly not an issue, but there's a reason why we had to do a _lot_ of wrangling w.r.t. the lock watch manager. Let's sync up about this offline.

File: lock-api/src/main/java/com/palantir/lock/client/TransactionStarter.java [hunk 14]
```diff
@@ -31,7 +31,7 @@
  * Callers of this class should use {@link #unlock(Set)} and {@link #refreshLockLeases(Set)} for returned lock tokens,
  * rather than directly calling delegate lock service.
  */
-final class TransactionStarter implements AutoCloseable {
+public final class TransactionStarter implements AutoCloseable {
```
Comment (Jolyon-S) [inline lock-api/src/main/java/com/palantir/lock/client/TransactionStarter.java:34]: same here

**Review by Jolyon-S — COMMENTED** [review]
This looks like an awesome step in the right direction - thanks for doing this! Got a number of comments, but most of them are minor stylistic comments rather than major design pieces.

I think my main concern is adding visibility to API, but it may be unfounded - let's discuss and see.

File: atlasdb-config/src/main/java/com/palantir/atlasdb/factory/DefaultLockAndTimestampServiceFactory.java (hunk 1 above)
Comment (gsheasby) [reply atlasdb-config/src/main/java/com/palantir/atlasdb/factory/DefaultLockAndTimestampServiceFactory.java:347]: So I _could_, but would also need to pass in `multiClientTimelockServiceSupplier` (and thus pass in two empty optionals from `IMTS`).

I'd lean against doing this, but let me know if you'd insist on it.

File: atlasdb-tests-shared/src/main/java/com/palantir/atlasdb/transaction/impl/AbstractSerializableTransactionTest.java (hunk 2 above)
Comment (gsheasby) [reply atlasdb-tests-shared/src/main/java/com/palantir/atlasdb/transaction/impl/AbstractSerializableTransactionTest.java:134]: Yes, it's necessary (otherwise we get a load of lock watch-related failures).

File: atlasdb-tests-shared/src/main/java/com/palantir/atlasdb/transaction/impl/TransactionTestSetup.java [hunk 15]
```diff
@@ -197,7 +196,7 @@ protected KeyValueService getKeyValueService() {
     }
 
     protected TransactionManager getManager() {
-        return tmManager.getLastRegisteredTransactionManager().orElseGet(this::createAndRegisterManager);
+        return createAndRegisterManager();
```
[the remark this answers is not in the document]
Comment (gsheasby) [reply atlasdb-tests-shared/src/main/java/com/palantir/atlasdb/transaction/impl/TransactionTestSetup.java:199]: rolled back.

File: atlasdb-tests-shared/src/main/java/com/palantir/timelock/paxos/DelegatingTimelockService.java (hunk 5 above)
Comment (gsheasby) [reply atlasdb-tests-shared/src/main/java/com/palantir/timelock/paxos/DelegatingTimelockService.java:55]: no, LegacyTimelockService still uses it.

File: atlasdb-tests-shared/src/main/java/com/palantir/timelock/paxos/DelegatingTimelockService.java (hunk 6 above)
Comment (gsheasby) [reply atlasdb-tests-shared/src/main/java/com/palantir/timelock/paxos/DelegatingTimelockService.java:60]: it was probably me trying to chase down some weird stuff in particular tests; I'll revert it and make sure that everything still works.

File: atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java (hunk 8 above)
Comment (gsheasby) [reply atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java:1211]: sadly there are two outputs (res and transactionTs); I did the best I could by extracting out `startTransactionWithWatches()`.

File: atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java (hunk 11 above)
Comment (gsheasby) [reply atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java:2144]: leftover from test-debugging pain; removed

File: atlasdb-tests-shared/src/test/java/com/palantir/timelock/paxos/InMemoryTimelockServicesTest.java (hunk 12 above)
Comment (gsheasby) [reply atlasdb-tests-shared/src/test/java/com/palantir/timelock/paxos/InMemoryTimelockServicesTest.java:89]: yeah, there's a separate fast forward test; for this one I just wanted to make sure they're monotonically increasing if we shuffle which methods to use.

File: timelock-agent/src/main/java/com/palantir/timelock/paxos/TimeLockAgent.java [hunk 16]
```diff
@@ -609,4 +611,9 @@ public void shutdown() {
         timestampStorage.close();
         sqliteDataSource.close();
     }
+
+    public ConjureTimelockService getConjureTimeLockService() {
+        return ConjureTimelockResource.jersey(
+                redirectRetryTargeter(), ns -> namespaces.get(ns).getTimelockService());
+    }
```
[the remark this answers is not in the document]
Comment (gsheasby) [reply timelock-agent/src/main/java/com/palantir/timelock/paxos/TimeLockAgent.java:618]: we don't 🎉

File: lock-api/src/main/java/com/palantir/lock/client/RemoteTimelockServiceAdapter.java [hunk 17]
```diff
@@ -41,14 +41,28 @@
...
     private static final SafeLogger log = SafeLoggerFactory.get(RemoteTimelockServiceAdapter.class);
 
-    private final NamespacedTimelockRpcClient rpcClient;
+    private final NamespacedTimelockTimestampClient rpcClient;
     private final NamespacedConjureTimelockService conjureTimelockService;
     private final LockLeaseService lockLeaseService;
     private final TransactionStarter transactionStarter;
     private final CommitTimestampGetter commitTimestampGetter;
 
+    // TODO(gs): factory method createInMemory?
```
Comment (Jolyon-S) [inline lock-api/src/main/java/com/palantir/lock/client/RemoteTimelockServiceAdapter.java:50]: I think we decided not to, at least for now? Down to you whether you want to leave this or not

**Review by Jolyon-S — APPROVED** [review]
Looks good to me!
